components/scream/src/dynamics/homme/physics_dynamics_remapper.hpp

```diff
@@ -238,167 +293,467 @@ do_bind_field (const int ifield, const field_type& src, const field_type& tgt)
+
+  view_dims.size = N;
+  for (int i=0; i<N-1; ++i) {
+    view_dims.dims[i] = field_dims[i];
+    num_values /= field_dims[i];
   }
+  view_dims.dims[N-1] = num_values;
 }
 
 template<typename RealType>
 void PhysicsDynamicsRemapper<RealType>::
-do_remap_fwd() const
+initialize_device_variables()
 {
-  using pack_type = ekat::Pack<RealType,SCREAM_PACK_SIZE>;
-  using small_pack_type = ekat::Pack<RealType,SCREAM_SMALL_PACK_SIZE>;
+  const int num_fields = m_phys.size();
+
+  phys_ptrs   = decltype(phys_ptrs)      ("phys_ptrs",   num_fields);
+  phys_dims   = decltype(phys_dims)      ("phys_dims",   num_fields);
+  phys_layout = decltype(phys_layout)    ("phys_layout", num_fields);
+
+  dyn_ptrs   = decltype(dyn_ptrs)   ("dyn_ptrs",   num_fields);
+  dyn_dims   = decltype(dyn_dims)   ("dyn_dims",   num_fields);
+  dyn_layout = decltype(dyn_layout) ("dyn_layout", num_fields);
+
+  has_parent          = decltype(has_parent)          ("has_parent",               num_fields);
+  pack_alloc_property = decltype(pack_alloc_property) ("phys_pack_alloc_property", num_fields);
+  is_state_field_dev  = decltype(is_state_field_dev)  ("is_state_field_dev",       num_fields);
+  time_levels         = decltype(time_levels)         ("time_levels",              1);
+
+  auto h_phys_ptrs   = Kokkos::create_mirror_view(phys_ptrs);
+  auto h_phys_layout = Kokkos::create_mirror_view(phys_layout);
+  auto h_phys_dims   = Kokkos::create_mirror_view(phys_dims);
+
+  auto h_dyn_ptrs   = Kokkos::create_mirror_view(dyn_ptrs);
+  auto h_dyn_layout = Kokkos::create_mirror_view(dyn_layout);
+  auto h_dyn_dims   = Kokkos::create_mirror_view(dyn_dims);
+
+  auto h_has_parent          = Kokkos::create_mirror_view(has_parent);
+  auto h_pack_alloc_property = Kokkos::create_mirror_view(pack_alloc_property);
+  auto h_is_state_field_dev  = Kokkos::create_mirror_view(is_state_field_dev);
+  auto h_time_levels         = Kokkos::create_mirror_view(time_levels);
 
   const auto& tl = Homme::Context::singleton().get<Homme::TimeLevel>();
 
-  const int num_fields = m_phys.size();
   for (int i=0; i<num_fields; ++i) {
     const auto& phys = m_phys[i];
     const auto& dyn  = m_dyn[i];
 
-    const int itl = m_is_state_field[i] ? tl.n0 : -1;
-
     const auto& ph = phys.get_header();
     const auto& dh = dyn.get_header();
 
+    const auto& phys_dim = ph.get_identifier().get_layout().dims();
+    const auto& dyn_dim  = dh.get_identifier().get_layout().dims();
+
+    // If field has a parent, then its view has been subviewed. We
+    // do not want to remmap subviews, only the view of the parent,
+    // so we store this and no other info for these fields, then
+    // skip these fields during the remap.
+    if (ph.get_parent().lock() != nullptr) {
+      EKAT_REQUIRE_MSG(dh.get_parent().lock() != nullptr,
+                       "Error! If physics field has parent,"
+                       "dynamics field must also have parent.");
+      h_has_parent(i) = true;
```

Comment:
Should we check that, if has_parent is true, then the parent was also registered in the remapper? I'm wondering whether we might not realize we're not remapping the subfield b/c the parent was never registered...

Reply:
Yeah, that's a good idea.